Use CoordsXY* on world/Footpath.cpp

tupaschoal opened the pull request:
I'm trying to get rid of non coords usage in Footpath. Some notes:

- Some places seemed like they could use a `CoordsXYRangedZD` with the `Direction`, but I was not sure if this should be done.
- Some other places could benefit from Map's `CoordsXYE`, would that be ok?
- The only left are `footpath_update_path_wide_flags(int32_t x, int32_t y);` which I couldn't tell for sure are `CoordsXY` and `footpath_get_coordinates_from_pos` + `footpath_bridge_get_info_from_pos` which would involve changing from altering parameter to returning coords.

Gymnasiast commented on src/openrct2/actions/LargeSceneryPlaceAction.hpp:
```diff
@@ -304,7 +304,7 @@ DEFINE_GAME_ACTION(LargeSceneryPlaceAction, GAME_COMMAND_PLACE_LARGE_SCENERY, La
             if (!(GetFlags() & GAME_COMMAND_FLAG_GHOST))
             {
-                footpath_remove_litter(curTile.x, curTile.y, zLow * 8);
+                footpath_remove_litter({ curTile, zLow * 8 });
```
```suggestion
                footpath_remove_litter({ curTile, zLow * COORDS_Z_STEP });
```

Gymnasiast commented on src/openrct2/actions/PlaceParkEntranceAction.hpp:
```diff
@@ -171,7 +171,7 @@ DEFINE_GAME_ACTION(PlaceParkEntranceAction, GAME_COMMAND_PLACE_PARK_ENTRANCE, Ga
             if (!(flags & GAME_COMMAND_FLAG_GHOST))
             {
-                footpath_connect_edges(entranceLoc.x, entranceLoc.y, newElement, 1);
+                footpath_connect_edges(entranceLoc, newElement, 1);
```
```suggestion
                footpath_connect_edges(entranceLoc, newElement, GAME_COMMAND_FLAG_APPLY);
```

Gymnasiast commented on src/openrct2/peep/GuestPathfinding.cpp:
```diff
@@ -142,13 +142,13 @@ static int32_t guest_surface_path_finding(Peep* peep)
     Direction randDirection = scenario_rand() & 3;
 
-    if (!fence_in_the_way(x, y, z, z + 4, randDirection))
+    if (!fence_in_the_way({ x, y, z * 8, (z * 8) + 4 * 8 }, randDirection))
```
Can you create a constexpr called `PATH_HEIGHT` and define it as 4 * COORDS_Z_STEP, then use that here?

```suggestion
    if (!fence_in_the_way({ x, y, z * COORDS_Z_STEP, (z * COORDS_Z_STEP) + PATH_HEIGHT }, randDirection))
```

Gymnasiast commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -696,25 +692,26 @@ static TileElement* footpath_get_element(int32_t x, int32_t y, int32_t z0, int32
-    int32_t z = tileElement->base_height;
-    TileElement* otherTileElement = footpath_get_element(x1, y1, z - 2, z, direction);
+    int32_t z = tileElement->GetBaseZ();
+    TileElement* otherTileElement = footpath_get_element({ shiftedFootpathPos, z - (2 * 8), z }, direction);
```
*cough* COORDS_Z_STEP *cough*

Gymnasiast commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1279,39 +1273,39 @@ static bool get_next_direction(int32_t edges, int32_t* direction)
 
     // Check if we are at edge of map
-    if (x < 32 || y < 32)
+    if (shiftedFootpathPos.x < 32 || shiftedFootpathPos.y < 32)
```
Use COORDS_XY_STEP.

Gymnasiast commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1279,39 +1273,39 @@ static bool get_next_direction(int32_t edges, int32_t* direction)
+            if (direction_reverse(slopeDirection) != shiftedFootpathPos.direction)
                 continue;
-            if (tileElement->base_height + 2 != z)
+            if (tileElement->GetBaseZ() + (2 * 8) != shiftedFootpathPos.z)
```
PATH_HEIGHT_STEP

Gymnasiast commented:
> Some places seemed like they could use a CoordsXYRangedZD with the Direction, but I was not sure if this should be done.

If they represent a thing together (e.g. an entrance, exit), then I would say yes.

> Some other places could benefit from Map's CoordsXYE, would that be ok?

Certainly, that's why it was created.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,51 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
         if (tileElement[1] == nullptr)
             continue;
 
         direction = (direction + 1) & 3;
```
direction_next

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,51 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
         tileElement[2]->AsPath()->SetCorners(tileElement[2]->AsPath()->GetCorners() | (1 << (direction)));
 
-        map_invalidate_element({ x, y }, tileElement[2]);
+        map_invalidate_element(currentPos, tileElement[2]);
```
This and subsequent calls to map_invalidate_element are calling with the wrong position. (note this is a bug in the current code)

tupaschoal replied:
I'm guessing this is because the coords here should be `footpathPos + delta` for each of their directions, correct? This is what I've used.

If so, shouldn't all the places doing `currentPos += delta` actually be doing `currentPos = footpathPos + delta`?

tupaschoal replied:
@duncanspumpkin can you weigh on this? If it's not clear I can explain it better.

tupaschoal replied:
Solving based on discussions on gitter.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,51 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
-        map_invalidate_element({ x, y }, tileElement[3]);
+        map_invalidate_element(currentPos, tileElement[3]);
 
         direction = (direction - 1) & 3;
```
direction_prev

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -731,13 +728,13 @@ static bool footpath_reconnect_queue_to_path(int32_t x, int32_t y, TileElement*
             otherTileElement->AsPath()->SetEdges(otherTileElement->AsPath()->GetEdges() | (1 << ((direction + 2) & 3)));
```
These direction+2's should be direction_reverse()

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -881,59 +877,59 @@ static void loc_6A6D7E(
 
 static void loc_6A6C85(
-    int32_t x, int32_t y, int32_t direction, TileElement* tileElement, int32_t flags, bool query,
+    const CoordsXY& mapPos, int32_t direction, TileElement* tileElement, int32_t flags, bool query,
```
mapPos is a bad name this is the `footpathPosition`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -772,31 +769,29 @@ static bool footpath_disconnect_queue_from_path(int32_t x, int32_t y, TileElemen
-    int32_t x = initialX + CoordsDirectionDelta[direction].x;
-    int32_t y = initialY + CoordsDirectionDelta[direction].y;
-    if (((gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) || gCheatsSandboxMode) && map_is_edge({ x, y }))
+    auto shiftedPos = CoordsXY{ pos } + CoordsDirectionDelta[pos.direction];
```
shiftedPos represents the coordinates of of the footpath/ride/entrance/mapEdge that is hopefully being connected to. `targetPos` might be a better name.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -772,31 +769,29 @@ static bool footpath_disconnect_queue_from_path(int32_t x, int32_t y, TileElemen
 static void loc_6A6D7E(
-    int32_t initialX, int32_t initialY, int32_t z, int32_t direction, TileElement* initialTileElement, int32_t flags,
-    bool query, rct_neighbour_list* neighbourList)
+    const CoordsXYZD& pos, TileElement* initialTileElement, int32_t flags, bool query, rct_neighbour_list* neighbourList)
```
pos is the initialTileElements position

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -881,59 +877,59 @@ static void loc_6A6D7E(
     }
 }
 
 static void loc_6A6C85(
```
Future refactoring job would be to change this into a simple check that validates if the direction should be fully checked with `loc_6A6D7E` and move the calling of `loc_6A6D7E` into the parent function.

tupaschoal replied:
I've noted on my personal notes, do you want me to add a `TODO` as well?

duncanspumpkin replied:
Yeah you could add a todo as it won't be for a while

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
 
     lastPathElement = nullptr;
     lastQueuePathElement = nullptr;
-    int32_t z = tileElement->base_height;
```
undo this change. The z coordinate is not part of the last path coordinates.

Gymnasiast replied:
Hm, are you sure we shouldn't save the z along with it? How will it handle paths above each other?

duncanspumpkin replied:
Its not part of the last path coordinate its a separate variable that is handled differently. Coupling it to last path would be confusing.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
-        tileElement = map_get_first_element_at({ x, y });
+        auto shiftedPos = footpathPos + CoordsDirectionDelta[direction];
```
`targetPos`

duncanspumpkin replied:
`targetQueuePos` = `curQueuePos` + ...

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
  */
 void footpath_chain_ride_queue(
-    ride_id_t rideIndex, int32_t entranceIndex, int32_t x, int32_t y, TileElement* tileElement, int32_t direction)
+    ride_id_t rideIndex, int32_t entranceIndex, const CoordsXY& footpathPos, TileElement* tileElement, int32_t direction)
```
Change footpathPos to `initialFootpathPos`. Rename `tileElement` to `initialTileElement` make it const.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
-    ride_id_t rideIndex, int32_t entranceIndex, int32_t x, int32_t y, TileElement* tileElement, int32_t direction)
+    ride_id_t rideIndex, int32_t entranceIndex, const CoordsXY& footpathPos, TileElement* tileElement, int32_t direction)
 {
     TileElement *lastPathElement, *lastQueuePathElement;
```
will also require `tileElement` set to the value of `initialTileElement`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
 {
     TileElement *lastPathElement, *lastQueuePathElement;
-    int32_t lastPathX = x, lastPathY = y, lastPathDirection = direction;
+    auto lastPath = CoordsXYZD{ footpathPos, tileElement->GetBaseZ(), static_cast<Direction>(direction) };
```
Create a new variable called `curQueuePos` initial value of `initialFootpathPos`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
                 if (tileElement->AsPath()->GetSlopeDirection() == direction)
                 {
-                    z += 2;
+                    lastPath.z += (2 * 8);
```
undo as above

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
-            lastPathX = x;
-            lastPathY = y;
-            lastPathDirection = direction;
+            lastPath.x = footpathPos.x;
```
value will be `curQueuePos`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1130,7 +1124,7 @@ void footpath_chain_ride_queue(
             tileElement->AsPath()->SetStationIndex(entranceIndex);
 
-            map_invalidate_element({ x, y }, tileElement);
+            map_invalidate_element(shiftedPos, tileElement);
```
add `curQueuePos = targetQueuePos;` pass `curQueuePos` into this function

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1045,32 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
  *  rct2: 0x006A742F
  */
 void footpath_chain_ride_queue(
```
The changes you have made will break this function. It might be easier to start again. It needs to be reworked this function as its a confusing mess. I've made some suggestions but they need to be thought through a bit more.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1279,39 +1273,39 @@ static bool get_next_direction(int32_t edges, int32_t* direction)
 
-    x += CoordsDirectionDelta[direction].x;
-    y += CoordsDirectionDelta[direction].y;
+    auto shiftedFootpathPos = CoordsXYZD{ CoordsXY{ footpathPos } + CoordsDirectionDelta[footpathPos.direction], footpathPos.z,
```
This is not a good name. `targetPos` while still vague would be better.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1279,39 +1273,39 @@ static bool get_next_direction(int32_t edges, int32_t* direction)
 static int32_t footpath_is_connected_to_map_edge_recurse(
-    int32_t x, int32_t y, int32_t z, int32_t direction, int32_t flags, int32_t level, int32_t distanceFromJunction,
-    int32_t junctionTolerance)
+    const CoordsXYZD& footpathPos, int32_t flags, int32_t level, int32_t distanceFromJunction, int32_t junctionTolerance)
```
I'm not convinced that direction should be part of footpathPos. It is used independently and represents the targetDirection and not the `footpathPos`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1396,26 +1392,27 @@ static int32_t footpath_is_connected_to_map_edge_recurse(
-int32_t footpath_is_connected_to_map_edge(int32_t x, int32_t y, int32_t z, int32_t direction, int32_t flags)
+int32_t footpath_is_connected_to_map_edge(const CoordsXYZD& footpathPos, int32_t flags)
 {
     flags |= (1 << 0);
```
Future job will be to use the `GAME_COMMAND_FLAGS`

tupaschoal replied:
I've added a `TODO` to track it.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1912,45 +1909,44 @@ void footpath_update_queue_entrance_banner(int32_t x, int32_t y, TileElement* ti
  */
-static void footpath_remove_edges_towards_here(
-    int32_t x, int32_t y, int32_t z, int32_t direction, TileElement* tileElement, bool isQueue)
+static void footpath_remove_edges_towards_here(const CoordsXYZD& footpathPos, TileElement* tileElement, bool isQueue)
```
Don't combine direction into the coordinate. Again this represents a direction not the footpathPos.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1912,45 +1909,44 @@ void footpath_update_queue_entrance_banner(int32_t x, int32_t y, TileElement* ti
-    x += CoordsDirectionDelta[direction].x;
-    y += CoordsDirectionDelta[direction].y;
+    Direction shiftedDirection = (footpathPos.direction + 1) & 3;
+    auto shiftedFootpathPos = CoordsXYZD{ CoordsXY{ footpathPos } + CoordsDirectionDelta[shiftedDirection], footpathPos.z,
```
This is a `targetFootpathPos` and the D value is not part of the target

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1912,45 +1909,44 @@ void footpath_update_queue_entrance_banner(int32_t x, int32_t y, TileElement* ti
         if (tileElement->GetType() != TILE_ELEMENT_TYPE_PATH)
             continue;
-        if (tileElement->base_height != z)
+        if (tileElement->GetBaseZ() != footpathPos.z)
```
use the `targetFootpathPos.z`

tupaschoal commented:
I've fixed all opened issues in their respective commits, for ease of reviewing. There is only one issue pending, which I need confirmation.

> For a lot of these functions direction can't be combined into the Coordinate as you will end up coupling two variables that are not related. As these functions are already confusing it will just make the issue worse.

There's none left.

> The changes you have made will break this function. It might be easier to start again. It needs to be reworked this function as its a confusing mess. I've made some suggestions but they need to be thought through a bit more.

This refers to `footpath_chain_ride_queue`. I've fixed all issues and moved it last in the commit order, in case we want to drop it.

duncanspumpkin commented on src/openrct2/peep/Staff.cpp:
```diff
@@ -727,10 +727,14 @@ static uint8_t staff_direction_surface(Peep* peep, uint8_t initialDirection)
 
-        if (fence_in_the_way(peep->next_x, peep->next_y, peep->next_z, peep->next_z + 4, direction))
+        if (fence_in_the_way(
+                { peep->next_x, peep->next_y, peep->next_z * COORDS_Z_STEP, (peep->next_z * COORDS_Z_STEP) + PATH_HEIGHT },
```
This isn't the path height it is the peep clearance height.

tupaschoal replied:
Add this constant to `Peep.h`

duncanspumpkin replied:
Yes

duncanspumpkin commented on src/openrct2/peep/Staff.cpp:
```diff
@@ -727,10 +727,14 @@ static uint8_t staff_direction_surface(Peep* peep, uint8_t initialDirection)
 
-        if (fence_in_the_way(peep->next_x, peep->next_y, peep->next_z, peep->next_z + 4, direction_reverse(direction)))
+        if (fence_in_the_way(
+                { peep->next_x, peep->next_y, peep->next_z * COORDS_Z_STEP, (peep->next_z * COORDS_Z_STEP) + PATH_HEIGHT },
```
as above

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -514,7 +514,7 @@ static TileElement* footpath_connect_corners_get_neighbour(int32_t x, int32_t y,
  *  rct2: 0x006A70EB
  */
-static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileElement* initialTileElement)
+static void footpath_connect_corners(const CoordsXY& footpathPos, TileElement* initialTileElement)
```
Can I be a pain and request this takes in a `PathElement*` instead of a `TileElement*`. It will reduce the AsPath() and mean the errors involved are higher up.

tupaschoal replied:
Done

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,50 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
-        int32_t x = initialX;
-        int32_t y = initialY;
-        int32_t direction = initialDirection;
+        auto currentPos = footpathPos;
```
The logic here has been changed. currentPos should be footpathPos + CoordsDirectionDelta[direction]

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,50 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
     {
-        int32_t x = initialX;
-        int32_t y = initialY;
-        int32_t direction = initialDirection;
```
I would keep this line in.

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,50 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
-        tileElement[2] = footpath_connect_corners_get_neighbour(x, y, z, (1 << direction_reverse(direction)));
+        auto directionNext = (initialDirection + 1) & 3;
```
I meant use the `direction_next()` function.

tupaschoal replied:
Argh, sure, my bad,

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -524,55 +524,50 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
-        map_invalidate_element({ x, y }, tileElement[3]);
+        directionNext = (directionNext + 1) & 3;
+        tileElement[3]->AsPath()->SetCorners(tileElement[3]->AsPath()->GetCorners() | (1 << (directionNext)));
+        map_invalidate_element(footpathPos + CoordsDirectionDelta[directionNext], tileElement[3]);
```
This is too confusing to follow. Instead of recording just the tileElement in the array record the position as well using a CoordsXYE struct then you don't need to recalculate the position. (Note: direction will still need to be calculated as I can't at this moment follow if it stays the same)

tupaschoal replied:
I stored a pair `std:;pair<PathElement*, CoordsXY>`, because I had to start storing `PathElement*` to comply with a suggestion above and they don't inherit from each other (thus we have to add some `reinterpret_cast`)

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -696,48 +691,51 @@ static TileElement* footpath_get_element(int32_t x, int32_t y, int32_t z0, int32
 
-    int32_t x1 = x + CoordsDirectionDelta[direction].x;
-    int32_t y1 = y + CoordsDirectionDelta[direction].y;
+    auto shiftedFootpathPos = footpathPos + CoordsDirectionDelta[direction];
```
`targetQueuePos` would be a nicer name

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -696,48 +691,51 @@ static TileElement* footpath_get_element(int32_t x, int32_t y, int32_t z0, int32
-    int32_t z = tileElement->base_height;
-    TileElement* otherTileElement = footpath_get_element(x1, y1, z - 2, z, direction);
+    int32_t z = tileElement->GetBaseZ();
+    TileElement* otherTileElement = footpath_get_element({ shiftedFootpathPos, z - LAND_HEIGHT_STEP, z }, direction);
```
`targetFootpathElement`. I would then make a new variable called `targetQueueElement` after confirming it is not a nullptr and that it is a queue. Then we can get rid of all of the `AsPath()`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1049,34 +1050,34 @@ void footpath_connect_edges(int32_t x, int32_t y, TileElement* tileElement, int3
             lastPathElement = tileElement;
-            lastPathX = x;
-            lastPathY = y;
-            lastPathDirection = direction;
```
I think you still need the `lastPathDirection`

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -1156,9 +1158,9 @@ void footpath_chain_ride_queue(
         {
             lastPathElement->AsPath()->SetHasQueueBanner(true);
-            lastPathElement->AsPath()->SetQueueBannerDirection(lastPathDirection); // set the ride sign direction
+            lastPathElement->AsPath()->SetQueueBannerDirection(direction); // set the ride sign direction
```
I think this needs to stay as `lastPathDirection`

tupaschoal commented:
I've fixed all the new issues. Please give me feedback on the one still open: https://github.com/OpenRCT2/OpenRCT2/pull/10515#discussion_r363987668

duncanspumpkin commented on src/openrct2/world/Footpath.cpp:
```diff
@@ -514,65 +515,65 @@ static TileElement* footpath_connect_corners_get_neighbour(int32_t x, int32_t y,
-        direction = initialDirection;
-        tileElement[0]->AsPath()->SetCorners(tileElement[0]->AsPath()->GetCorners() | (1 << (direction)));
-        map_invalidate_element({ x, y }, tileElement[0]);
+        direction = direction_prev(direction);
```
Its the same value but can you make this the same as the original i.e. set it to initialDirection.

Gymnasiast commented on src/openrct2/Editor.cpp:
```diff
@@ -500,12 +500,9 @@ namespace Editor
             int32_t direction = direction_reverse(parkEntrance.direction);
 
-            switch (footpath_is_connected_to_map_edge(x, y, z, direction, 0))
+            switch (footpath_is_connected_to_map_edge(parkEntrance, direction, 0))
```
CoordsXYZD would be preferable here, I'd say.

duncanspumpkin replied:
The direction represents where you want to check where as the xyz represents the footpath tile you are looking from. They shouldn't be tied together.

tupaschoal replied:
Not changing anything then

Gymnasiast commented on src/openrct2/peep/GuestPathfinding.cpp:
```diff
@@ -142,13 +142,13 @@ static int32_t guest_surface_path_finding(Peep* peep)
     int16_t z = peep->next_z;
```
Wouldn't it be better to just create a CoordsXYRangedZ and feed that into the functions? That should make the code easier to read.

Alternatively, making `z` use big coordinates would also clean up a lot already.

tupaschoal commented:
Should be alright now.
